ipaserver/plugins/cert.py

```diff
@@ -787,6 +800,62 @@ def execute(self, csr, all=False, raw=False, **kw):
         )
 
 
+def _dns_name_matches_principal(name, principal, principal_obj):
+    """
+    Ensure that a DNS name matches the given principal.
+
+    :param name: The DNS name to match
+    :param principal: The subject ``Principal``
+    :param principal_obj: The subject principal's LDAP object
+    :return: True if name matches, otherwise False
+
+    """
+    for alias in _collect_principal_aliases(principal_obj):
+        # we can only compare them if both subject principal and
+        # the alias are service or host principals
+        if not (alias.is_service and principal.is_service):
+            continue
+
+        # ignore aliases with different realm or service name from
+        # subject principal
+        if alias.realm != principal.realm:
+            continue
+        if alias.service_name != principal.service_name:
+            continue
+
+        # now compare DNS name to alias hostname
+        if name.lower() == alias.hostname.lower():
+            return True  # we have a match
+
+    return False
+
+
+def _principal_name_matches_principal(name, principal_obj):
+    """
+    Ensure that a stringy principal name (e.g. from UPN
+    or KRB5PrincipalName OtherName) matches the given principal.
+
+    """
+    try:
+        principal = kerberos.Principal(name)
+    except ValueError:
+        return False
+
+    return principal in _collect_principal_aliases(principal_obj)
+
+
+def _collect_principal_aliases(principal_obj):
+    """
+    Return all Kerberos ``Principal`` aliases for given principal
+    (including the canonical name).
+
+    """
+    aliases = set(principal_obj.get('krbprincipalname', []))
+    for alias in principal_obj.get('krbcanonicalname', []):
```

Comment:
OK, thanks.  I've removed function `_collect_princpial_aliases()` and just replaced with `principal_obj.get('krbprincipalname', [])` at call sites.